Stdlib functional iterators

c-cube opened the pull request:
See #635 for previous discussion and design.

This new proposal uses a purely functional type of delayed lists, as suggested by @fpottier
and already implemented in Batteries. The module is named `Seq`.

```ocaml
type 'a node =
  | Nil
  | Cons of 'a * 'a t

and 'a t = unit -> 'a node
```

The type is not structural, sadly (using `type 'a t = unit -> ('a * 'a t) option` does not pass the acyclicity check), but is very simple to use, and has decent performances (see previous PR). Ping @drup @alainfrisch @gasche @jakemcarthur @paurkedal for more discussions.

About the choice of `unit -> 'a node` rather than `'a node lazy_t`: it is much faster if the iterator is traversed once, and it entails fewer memory leaks (e.g. for `range 0 1_000_000`, there will be no memoization). One can still write a memoization combinator (I wrote one for `gen` which is very efficient, using an underlying buffer).

c-cube committed:
add `Seq` module, expose iterator conversions in most containers

paurkedal commented on stdlib/StdlibModules:
```diff
@@ -42,6 +42,7 @@ STDLIB_MODULES=\
   hashtbl \
   int32 \
   int64 \
+  iter \
```
Should be seq, right?

paurkedal commented on stdlib/bytes.ml:
```diff
@@ -327,3 +327,41 @@ let lowercase s = map Char.lowercase s
+
+let of_seq i =
+  let n = ref 0 in
+  let buf = ref (make 256 '\000') in
```
Would be simpler to use `Buffer`, since geometric resizing is used anyway.  (But good you avoid double iteration, since we don't know the complexity of traversing the input.)

c-cube replied:
Yes, relying on `Buffer` would be more convenient, but it would create circular dependencies since `Buffer` already depends on `Bytes`.

paurkedal replied:
Of course. Your partial reimplementation is quite lean anyway.

paurkedal commented on stdlib/hashtbl.ml:
```diff
@@ -354,6 +354,35 @@ let stats h =
     max_bucket_length = mbl;
     bucket_histogram = histo }
 
+(** {6 Seq.tors} *)
```
Search & replace error in heading?

paurkedal commented on stdlib/map.mli:
```diff
@@ -283,7 +283,24 @@ module type S =
+        in ascending order, from key [k] or above.
+        @since NEXT_RELEASE *)
+
+    val add_seq : 'a t -> (key * 'a) Seq.t -> 'a t
```
`val add_seq : (key * 'a) Seq.t -> 'a t -> 'a t` for consistency with `add` and to make the partial application composable.

c-cube replied:
Indeed, will do.

paurkedal commented on stdlib/set.mli:
```diff
@@ -258,6 +258,25 @@ module type S =
+    (** Iterate on the whole set, in ascending order
+        @since NEXT_RELEASE *)
+
+    val add_seq : t -> elt Seq.t -> t
```
`add_seq : elt Seq.t -> t -> t`, cf comment on `Map.add_seq`.

c-cube replied:
fixed in last commit

c-cube committed:
small typo

paurkedal reviewed (commented):
I generally think this approach looks good; since this is meant as a generic way to get out data from containers, I prefer it simple and performant rather than abstract (which I agree would be needed) and featureful. Though, I don't know how the performance compares to the previous PR.

c-cube committed:
typo

c-cube commented:
There were some basic performance comparisons in #635. As far as `map`, `flat_map` and the likes are concerned, @Drup made some benchmarks a while ago that put the various types discussed previously roughly in the same order, except for `('a -> unit) -> unit` that is much faster on some of the operations it supports.

c-cube committed:
change order of arguments for `{Map,Set}.add_seq`

c-cube committed:
watch for max string length in `Bytes.of_seq`

paurkedal commented on stdlib/bytes.ml:
```diff
@@ -327,3 +327,43 @@ let lowercase s = map Char.lowercase s
+  let resize () =
+    (* resize *)
+    let new_len = min (2 * length !buf) Sys.max_string_length in
+    if length !buf = new_len then failwith "Bytes.of_seq: cannot grow bytes";
```
Slightly simpler:
```ocaml
    let new_len = 2 * length !buf in
    if new_len > Sys.max_string_length then failwith "Bytes.of_seq: cannot grow bytes";
```
(Both assuming `max_string_length < 2 * max_int`, which is also assumed in `Buffer`.)

alainfrisch commented on stdlib/array.ml:
```diff
@@ -292,3 +292,40 @@ let stable_sort cmp a =
+      in
+      fill (len-1) tl
+
+let of_seq i =
```
This can be improved later, but this function might deserve a more efficient implementation which does not build the intermediate list at least for small enough arrays (e.g. keeping the first N values on the stack using a non-tail rec loop and creating the array while unwinding the recursion).  Even for larger arrays, it might be more efficient to keep the intermediate values directly in arrays (either a resizing buffer, or a list of chunks).  To be checked.

c-cube replied:
Why not, will consider it indeed.

alainfrisch commented on stdlib/buffer.ml:
```diff
@@ -194,3 +194,31 @@ let truncate b len =
+
+let to_seq b =
+  let rec aux i () =
+    if i = b.position then Seq.Nil
```
Because of `Buffer.{clear, reset, truncate}`, you should probably test for `i >= b.position` instead.  (Even if mutation during iteration is specified as being undefined...)

c-cube replied:
will do.

alainfrisch commented on stdlib/buffer.ml:
```diff
@@ -194,3 +194,31 @@ let truncate b len =
+
+let to_seqi b =
+  let rec aux i () =
+    if i = b.position then Seq.Nil
```
Same as above.

alainfrisch commented on stdlib/array.mli:
```diff
@@ -256,6 +256,20 @@ val fast_sort : ('a -> 'a -> int) -> 'a array -> unit
+(** {6 Iterators} *)
+
+val to_seq : 'a array -> 'a Seq.t
+(** Iterate on the array, in increasing order
```
You should specify that mutation during iteration is not specified (as you did for Buffer).  Or, alternatively specify the behavior as you did for Bytes.

alainfrisch commented on stdlib/hashtbl.ml:
```diff
@@ -377,21 +406,32 @@ module type S =
-    val add: 'a t -> key -> 'a -> unit
-    val remove: 'a t -> key -> unit
-    val find: 'a t -> key -> 'a
+    val copy : 'a t -> 'a t
```
No reason for these whitespace changes (and the recommended style would rather be the one without the whitespace).

alainfrisch commented on stdlib/hashtbl.ml:
```diff
@@ -377,21 +406,32 @@ module type S =
     val stats: 'a t -> statistics
   end
 
+module type FULL =
```
Why do we need this distinction between S and FULL?

c-cube replied:
that was because of the sharing of `S` with some ephemeron table, but I removed the sharing, so this has no reason to exist anymore indeed.

alainfrisch replied:
@c-cube Can you address this point?

alainfrisch commented on stdlib/hashtbl.mli:
```diff
@@ -215,6 +215,36 @@ val stats : ('a, 'b) t -> statistics
    buckets by size.
    @since 4.00.0 *)
 
+(** {6 Seq.iterators} *)
```
Typo?  (also in the doc strings below)

alainfrisch commented on stdlib/map.mli:
```diff
@@ -283,7 +283,24 @@ module type S =
+    (** Iterate on the whole map, in ascending order
+        @since NEXT_RELEASE *)
+
+    val to_seq_at : key -> 'a t -> (key * 'a) Seq.t
```
I'm not very fond of the name.  `to_seq_starting_from` or `to_seq_from`, perhaps?

paurkedal replied:
I agree `to_seq_from` is more descriptive than `to_seq_at`.  But we could also incorporate more general slicing in the main sequence extractor,
```ocaml
val to_seq : ?first: (key -> bool) -> ?last: (key -> bool) -> 'a t -> (key * 'a) Seq.t
```
borrowing the semantics of `find_first` and `find_last`. While the end-of-sequence can be adjusted with a sequence combinator, direct support for it has the advantage of only holding on to the part of the container it needs plus `O(log(n)`.  **Corr.**: There is no such guarantee on space complexity here, since the sequence may contain a single element greater than the top-level node. The should be a gain in expected complexity for small sequence lengths.

alainfrisch commented on stdlib/seq.mli:
```diff
@@ -0,0 +1,40 @@
+
+val fold_left : ('a -> 'b -> 'a) -> 'a -> 'b t -> 'a
+
+val iter : ('a -> unit) -> 'a t -> unit
```
What about providing a function that memoize the iteration?

c-cube replied:
That is something non-trivial to do efficiently, in my experience. Using `lazy` is relatively slow, so I have something better (a kind of append-only buffer) in the `gen` library that could be used as a building block for `memoize`. However, this is imho the kind of function that might be more suited to a 3rd party library on opam (which could provide many more combinators and evolve faster than the stdlib).

alainfrisch replied:
Fair enough.  It might be worth adding a note to seq.mli stating that the generator must be prepared to be used in a non-linear way, so that people don't expose an overly naive generator iterating over lines of a text file.

alainfrisch commented:
I'm a bit concerned by the addition of a module to the stdlib with such a short name.  One way to avoid conflicts would be to add it as a submodule of Pervasives.  Another would be to use a longer name (CamlSeq) and expose an alias `module Seq = CamlSeq` in Pervasives.  (There probably requires some tweaks to the build system.)

shindere commented:
How about Sequence?

alainfrisch commented:
I wouldn't bet that sequence.ml is not used by any OCaml project (and actually, there is at least https://github.com/c-cube/sequence/blob/4e2595df8983f64cb094334f8b4e2851e3f837fa/src/sequence.ml).

c-cube commented:
Indeed, I have been using `Sequence` as a toplevel module, it's in several projects. Maybe the `Caml_seq` + alias would be better (although I'm afraid when a maintainer mentions "tweaks to the build system").

ghost commented:
I started working on a patch to prefix all the stdlib modules and use module aliases to get the short names back. It's all working, except for a few tests where types are not printed properly and some ocamldoc errors. I should be able to finish the patch tomorrow or next week, then we'll be free to add new modules to the stdlib without guilt. The branch is here: https://github.com/diml/ocaml/tree/prefix-the-stdlib

c-cube commented:
is it worth the trouble updating this PR now, or should I wait for #1010 to be merged?

c-cube commented on stdlib/ephemeron.ml:
```diff
@@ -21,7 +21,24 @@ module type SeededS = sig
+  val filter_map_inplace: (key -> 'a -> 'a option) -> 'a t -> unit
+  val fold : (key -> 'a -> 'b -> 'b) -> 'a t -> 'b -> 'b
+  val length : 'a t -> int
+  val stats: 'a t -> Hashtbl.statistics
```
I'm not even sure what it means to iterate on a weak hashtable… I will certainly not do that before #1010 is merged, my quota of motivation is running dry right now.

alainfrisch replied:
@bobot Do you agree with delaying the addition of Seq-related functions to ephemerons?

dorafmon commented on stdlib/seq.mli:
```diff
@@ -0,0 +1,57 @@
+
+val return : 'a -> 'a t
+
+val map : ('a -> 'b) -> 'a t -> 'b t
```
should we also have a `>>=`?

paurkedal replied:
Wouldn't that be the same as `flat_map` found below?

dorafmon commented on stdlib/seq.mli:
```diff
@@ -0,0 +1,57 @@
+
+val iter : ('a -> unit) -> 'a t -> unit
+
+val memo : 'a t -> 'a t
```
Should we have an imperative `yield` function?
```ocaml
let yield (x:' a Seq.t) : ('a option * 'a Sequence.t) = 
  let open Seq in match x () with 
  | Nil -> (None, empty)
  | Cons (a, r) -> (Some a, r)
```

c-cube replied:
you mean a memoizing `peek` or something like that? I don't really get what `yield` would be used for.

besides, until this has chances of being merged (i.e. #1010 merged?) I won't spend any more time on that PR.

dorafmon replied:
Hi, sorry for the late reply, it is more like [`next` in `Core.Sequence`](https://ocaml.janestreet.com/ocaml-core/latest/doc/core_kernel/Core_kernel/Sequence.mod/#/next.val). But we can wait until this PR is merged and then consider this function as a useful extension.  I personally used it a lot. Not sure how others feel.

ubsan commented:
I really dislike the impure nature of iterators implemented with this approach - I'd much prefer a "state and function" approach to iteration. Has something like that been considered? Why has it been rejected?

(see https://github.com/ubsan/pred/blob/master/pred/iter.ml for an example)

infinity0 commented:
nicole mazzuca:
> I really dislike the impure nature of iterators implemented with this approach - I'd much prefer a "state and function" approach to iteration. Has something like that been considered? Why has it been rejected?
> 
> (see https://github.com/ubsan/pred/blob/master/pred/iter.ml for an example)
> 

In case no-one answered this - I believe the point is that the state is implicitly part of the closures returned by subsequent calls to the first closure. When iterating it's generally a bit pointless to have the consumer deal with this state themselves, they don't care about it and only care about the output values. The producer has to track it anyway, so they might as well hide it from the consumer.

In summary, this is more like Haskell's lazy lists rather than a state monad. Of course nothing prevents you from implementing an iterator as a state monad, but using them would be more awkward.

strega-nil commented:
@infinity0 yeah, that's why I deleted my comment. I was thinking of them wrong.

c-cube committed:
wip: make it build again

alainfrisch commented:
The module can now be named Seq or Sequence.  @c-cube Would you find the motivation to resurrect the proposal?

c-cube commented:
It's named `Seq` currently, and I'll rebase it along #640 .

dbuenzli commented:
In the spirit of keeping the stdlib lightweight and to the point. I still don't see the advantage of this approach over the structural `iter` and `fold` functions that data structures usually provide.

Sure you don't get inversion of control with `Seq` but since this approach is not able in general to compose with IO end points I find the added flexibility to be of limited use (and wonder how many times we will have to explain to programmers that they should not hold on ressources in `Cons`'s continuation, cf. Haskell's lazy IO).

c-cube commented:
Do we really have to debate this over and over again? The stdlib is perceived by most people as *too* lightweight…

Still: just use this in conjunction with #640 and you're good to go w.r.t. resource handling. That's what I do whenever I open a file. I don't care much about IO iterators anyway, as IO are typically 1 or 2% of my code (ignoring printing on stdout), and I just use Lwt for system stuff.

c-cube commented:
Locally I have a seemingly unrelated build error in `tools/make_opcodes` (which doesn't find `Stdlib__sys` for some reason).

dbuenzli commented:
(This message has links comments of #635 which are hidden and that you will need to unfold to access)

> Do we really have to debate this over and over again? 

Yes. 

> The stdlib is perceived by most people as too lightweight…

Perception is a matter of communication. If people could stop shouting and parroting everywhere that OCaml needs alternate standard libraries (which in the long run have done more harm than good in my opinion) and that OCaml's standard library is only what is needed to implement the compiler that perception could certainly be changed; especially now that the dev team has made it clear that it was willing to let the stdlib evolve.

I don't think the stdlib is too lightweight. Some things are obviously missing, some things do not compose well and some things are not consistent enough all of which are fixable in the long run. However I don't think what you propose here is essential. Reading [your justification](https://github.com/ocaml/ocaml/pull/635#issuecomment-2291738459) for this data structure, I personally don't find it compelling enough for inclusion as a useful pervasive iterating structure given that it's not able to cope with resources in a satisfactory manner.

> Still: just use this in conjunction with #640 and you're good to go w.r.t. resource handling.

No you can't use that unless you read all your data in memory before (like you do [here](https://github.com/ocaml/ocaml/pull/635#issuecomment-230276588)). In many cases this is not desirable or even possible. The result is that people will write simple programs like you did there and suddenly wonder why they don't scale and consume too much memory.

bluddy commented:
I think the discussion in #635 is more than sufficient. It doesn't seem like @dbuenzli can ever be convinced of this point -- there will always be some possible objection.

rgrinberg commented:
What were the reasons for not choosing the implementation of Sequence as done in base? (the one by @jakemcarthur). I think that one is the most convincing one, and being able to write code generator style is extremely convenient (just look at the implementation of `split_full_gen` in ocaml-re for the alternative).

I'm also very happy to see such an iterator added to the stdlib. I understand that it's not applicable in every situation (what is?), but I found it to be quite useful in many situations. For example, splitting a string, or returning all the matches in the ocaml-re library would be much better handled by returning an iterator over a list. Consuming a `fold` or an `iter` function is not good enough and requires the user to raise an exception for early termination. I think that this is style is as brittle as any other misuse of iterators.

Therefore, I think that communication efforts to end users are better expanded on describing the limitations of this approach, rather than trying to convince them that they don't need. The fact that a variant of this iterator is present in all 3 standard libraries means that users are already convinced of its usefulness. The stdlib stepping in and standardizing what's an already widespread pattern seems like a net positive.

Also, if we have effects won't we be able to make this iterator safer by forcing purity on the iterator continuation?

alainfrisch commented:
I confirm the build issue.  Can be reproduced on the branch (after the failure), but not on trunk:
````
$ cat foo.ml
$ boot/ocamlrun boot/ocamlc -nostdlib -I boot foo.ml -o foo.exe
File "foo.ml", line 1:
Error: Required module `Stdlib__sys' is unavailable
````
And I checked that boot/ocamlc is the same as on trunk.

alainfrisch commented:
The problem is perhaps that Bytes is linked in stdlib.cma before Sys, but refers to it.

dbuenzli commented:
> The fact that a variant of this iterator is present in all 3 standard libraries means that users are already convinced of its usefulness.

What a funny argument. It's usage, not definition that matters. These large bureaucratic extended libraries tend to pick up "me too" bloat. On the many complaints I have seen over the years on the stdlib I don't think missing such a structure was a prevalent one (if ever mentioned).

> Also, if we have effects won't we be able to make this iterator safer by forcing purity on the iterator continuation?

Unlikely, the stdlib is rightfully quite stringent on backward compatibility. Also as mentioned in #635 effects might anyways bring new tools to the trade to effectively solve the problem in a satisfactory manner (another reason why I think it's not a good idea to settle on that one now).

Anyways I'll stop arguing now, my points have been made.

alainfrisch commented:
Hypothesis confirmed: you cannot refer to Sys.max_string_length from Bytes, or you need to link Sys before Bytes (in stdlib/Makefile).

c-cube commented:
> These large bureaucratic extended libraries

Sorry, but this is really making my blood boil :angry: . I don't consider containers more bloated or bureaucratic than your own bunch of modules, so please stick to the facts.

alainfrisch committed:
Fix dependency

Sys needs to be linked before Bytes in stdlib.

dbuenzli commented:
> Sorry, but this is really making my blood boil 😠 . I don't consider containers more bloated or bureaucratic than your own bunch of modules, so please stick to the facts.

Sorry @c-cube I hadn't this one in my mind. (I always thought containers was a data structure library rather than an alternate stdlib).

c-cube commented:
@dbuenzli ah I see. Then I'm not sure which one is the 3rd "extended library". Containers is really an extension of the stdlib, these days.

alainfrisch committed:
Update threads/stdlib.ml

dbuenzli commented:
Let's not get OT but nowadays there's at least `extlib`, `batteries`, `core`, `core_kernel`, `base` and apparently `containers` so I'm not sure three is an accurate number.

alainfrisch committed:
Update stdlib_no_prefixed/.depend

alainfrisch commented:
Ok, I pushed some commits to the PR branch to fix the build.

alainfrisch commented:
The build succeeds, but the tests fail.  @c-cube Can you look at it?

alainfrisch commented:
Ping @c-cube

c-cube committed:
fix inconsistencies with label modules

c-cube committed:
update testsuite to work with seq

c-cube committed:
update change file

c-cube committed:
small change in `Hashtbl.to_seq`, capturing only the underlying array

c-cube commented:
pong @alainfrisch 

(edit: added a few tests, fixed the existing ones)

alainfrisch commented on stdlib/seq.mli:
```diff
@@ -0,0 +1,40 @@
+  | Nil
+  | Cons of 'a * 'a t
+
+and 'a t = unit -> 'a node
```
Can you add some documentation to the module?  The general idea ("lazy lists", possibly infinite; should still be deterministic; etc), and a short text for each component.

Some other functions would make sense to add (mapi, map2, exists, for_all, etc).  Do you prefer to keep such discussions for later?  (Having them in this PR take, overall, less energy than adding them later, but I understand you'd prefer this PR to be merged sooner than later.)

c-cube replied:
Make your pick in https://github.com/c-cube/oseq/ :wink: 
I'll add a bit of documentation, but would really like to have this merged soon.

bluddy replied:
I think having these advanced functions later makes sense. The key thing is that we get access to an iterator that can only be made with the ability to access the internals of the stdlib data structures. Specific advanced functions can easily be added, but there's a possibility of some heavy bikeshedding at that point. Now that this mechanism exists, libraries can also take advantage of it.

c-cube committed:
add some documentation to seq.mli

c-cube committed:
revert to good ol' module type names for hashtables

c-cube committed:
fix test

c-cube commented:
updated @alainfrisch

alainfrisch commented on stdlib/seq.mli:
```diff
@@ -15,26 +15,63 @@
+(** The type of delayed lists containing elements of type ['a].
+    Note that the concrete list node ['a node] is delayed under a closure,
+    not a [lazy] block, which means it might be recomputed every time
+    we access it. *)
```
In the stdlib, doc comments come after declarations.  Could you follow this style?

alainfrisch commented:
I think there had been serious benchmarking between various implementations of lazy lists/iterators, but I cannot find immediately the comparison between the current proposal, and other existing similar types in Containers/Core, or even list itself.  Do you have such results available?  Basically, users will often be faced with the choice between Seq.t, list, or another types, and providing some hints about the respective performance would be useful.

c-cube committed:
change style of comments in seq.mli

c-cube commented:
@alainfrisch for the comments, done. For the perf comparison there's https://github.com/c-cube/iterators_bench/ . However I would **REALLY** like to see this PR merged now because I won't have the patience to do this rebase/update/scratch my head for another year :scream:

alainfrisch commented:
Frankly, I appreciate your heroic efforts to move this forward, I understand the process can be frustrating, and I'm trying to help, but this is an important change to the stdlib and it deserves proper review and polishing.  Btw, some review comments from last year remain unaddressed.

c-cube committed:
follow some demands in review of GPR #1002

c-cube commented:
Right. Apologies, I'm not at the top today. I've just udpated the code following some reviews, but might have missed some of them, would you be so kind as to indicate which ones are missing, if any?

alainfrisch commented on stdlib/hashtbl.mli:
```diff
@@ -296,8 +326,26 @@ module type S =
+    (** @since NEXT_RELEASE *)
   end
-(** The output signature of the functor {!Hashtbl.Make}. *)
+(** The core output signature of the functor {!Hashtbl.Make}. *)
```
This 'core' is no longer relevant.

alainfrisch commented on stdlib/hashtbl.mli:
```diff
@@ -352,8 +400,26 @@ module type SeededS =
+    (** @since NEXT_RELEASE *)
   end
-(** The output signature of the functor {!Hashtbl.MakeSeeded}.
+(** The core output signature of the functor {!Hashtbl.MakeSeeded}.
```
Same as above.

alainfrisch commented on Changes:
```diff
@@ -50,6 +50,10 @@ Working version
 
 ### Standard library:
 
+- GPR #1002: add a new `Seq` module defining a list-of-thunks style iterator.
```
Spurious whitespace after GPR.

alainfrisch commented:
LGTM, modulo a few tiny inline comments.  We will want to add more functions and optimize the implementation of current ones but this can be done later.

I will merge later this week unless someone objects to it.

@c-cube Can you replace NEXT_RELEASE with 4.07?

gasche commented:
Congratulations on getting this far!

What is (roughly) the amount of test coverage for the new functions? I only re-skimmed the patch, but I see tests for `Seq.filter`. Is the idea to eventually get more? (I think that basic tests for the various `of_seq`/`to_seq` functions for example could be interesting.)

c-cube committed:
some fixes for #1002

c-cube commented:
I'll do the ephemeron changes and that should be good, then. Thank you @alainfrisch!

c-cube committed:
add Seq-related functions to Ephemeron

c-cube commented:
I added the basics for `Ephemeron`, following @bobot's advice. @gasche I added a few tests for `Set` and `Map`, and adding basic `{to_seq,of_seq}` hashtbl tests right now.

c-cube committed:
add some comments on `Hashtbl.of_seq`

c-cube committed:
add more tests for `Hashtbl.{to,of}_seq`

c-cube committed:
fix bug in `Ephemeron.to_seq`

alainfrisch committed:
Merge branch 'trunk' into stdlib-iterators

alainfrisch commented:
I fixed a conflict in Makefile, now waiting for CI green light before merging.  I second @gasche's suggestion to add more tests at some point, though.

alainfrisch committed:
Update Changes

alainfrisch commented:
Merged, just in time for the week-end!  (With a squash merge, since the history did not seem so useful.)

c-cube commented:
Thank you @alainfrisch !! :grin: 
Now I think I should write a compatibility package for the `'a Seq.t` type, and add more tests.

alainfrisch commented:
(@c-cube It occurred to me that such contribution would probably require a CLA.  Have you already signed one?)

c-cube commented:
I don't think I have, but this was written on my free time, so there should be not problem. Where should I sign the CLA?

alainfrisch commented:
https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#contributor-license-agreement

c-cube commented:
There isn't an online form for the CLA? :confused: 
This is going to take time, I don't have a printer…

nojb commented:
Question: why use `Hashtbl.replace` in the implementation of `Hashtbl.of_seq`, instead of `Hashtbl.add` ?

Context: In #2236 we define `Hashtbl.of_list` and it seems natural to use `Hashtbl.add` to implement it (instead of `Hashtbl.replace`) but it also needs to be compatible with `Hashtbl.of_seq`.

damiendoligez commented:
@c-cube : Do you have a comment on this latest question?

c-cube commented:
That was a choice I made to prevent duplicates from making tables too big, but really there could have been more discussion on that particular point, I think. It would be useful to provide both, as there are good use cases for both add and replace semantics imho.
